iothub_client/tests/longhaul_resource_cleanup/longhaul_resource_cleanup.c

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/platform.h"
+#include "azure_c_shared_utility/threadapi.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
+#include "azure_c_shared_utility/shared_util_options.h"
+#include "iothub_device_client.h"
+#include "iothub_client_options.h"
+#include "iothub_message.h"
+#include "iothub_registrymanager.h"
+#include "iothubtransportamqp.h"
+#include "iothub_account.h"
+#include "iothubtest.h"
+#include "parson.h"
+#include "../common_longhaul/iothub_client_common_longhaul.h"
+
+#ifdef SET_TRUSTED_CERT_IN_SAMPLES
+#include "certs.h"
+#endif // SET_TRUSTED_CERT_IN_SAMPLES
+
+const int five_days = (60 * 60 * 24) * 5;
+const int year_2020 = 1577865600;
+static const char SAS_DEVICE_PREFIX_FMT[] = "csdk_e2eDevice_sas_j_please_delete_";
+static const char X509_DEVICE_PREFIX_FMT[] = "csdk_e2eDevice_x509_j_please_delete_";
+
+typedef struct ENUM_CONTEXT_TAG {
+    IOTHUB_REGISTRYMANAGER_HANDLE registryManagerHandle;
+    IOTHUB_SERVICE_CLIENT_DEVICE_TWIN_HANDLE deviceTwinHandle;
+} ENUM_CONTEXT;
+
+void deviceItemEnum(const void* item, const void* action_context, bool* continue_processing)
+{
+    const ENUM_CONTEXT* context = (const ENUM_CONTEXT*) action_context;
+    const IOTHUB_MODULE* device = (const IOTHUB_MODULE*) item;
+    if (strncmp(device->deviceId, SAS_DEVICE_PREFIX_FMT, sizeof(SAS_DEVICE_PREFIX_FMT) - 1) == 0 || 
+        strncmp(device->deviceId, X509_DEVICE_PREFIX_FMT, sizeof(X509_DEVICE_PREFIX_FMT) - 1) == 0)
+    {
+        char* twin = IoTHubDeviceTwin_GetTwin(context->deviceTwinHandle, device->deviceId);
+        if (twin)
+        {
+            JSON_Object* root_object = NULL;
```

Comment:
Are any of these variables dynamically allocated? If so, I don't see any frees afterwards.

Reply:
it's freed by json_value_free(root_value);

Reply:
see https://github.com/kgabis/parson